flutter: Add update script and fix version

FlafyDev opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This PR adds an update script that updates nixpkgs for a new version of flutter.
This PR also:
- fixes `flutter doctor` and `flutter --version` not reporting the correct version by adding a `flutter.version.json` file.
- modifies the structure for the Flutter package in Nixpkgs for multiple Flutter versions.
- updates flutter from 3.13.8 to 3.16.7

Fixes #272977

@NixOS/flutter 
@mkg20001 
@maximoffua

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

FlafyDev committed:
flutter: Add version json file

Needed for flutter doctor and flutter --version

Co-authored-by: Sergii Maksymov <ua.maximoff@gmail.com>

FlafyDev committed:
flutter: Rename platform to flutterPlatform

This helps avoid the ambiguity between Flutter platforms like Linux, IOS, and Android and platforms like linux-x64

FlafyDev committed:
flutter: Make fetch-artifects.nix independent from the host's platform

This is useful for an update script

FlafyDev commented:
Result of `nixpkgs-review pr 276050` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>localsend</li>
    <li>yubioath-flutter</li>
  </ul>
</details>
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>expidus.file-manager</li>
    <li>firmware-updater</li>
    <li>fluffychat</li>
    <li>flutter</li>
    <li>flutter-unwrapped</li>
    <li>hover</li>
    <li>intiface-central</li>
  </ul>
</details>

FlafyDev commented:
I can't build localsend and yubioath-flutter on master. so it's not this PR that broke them

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3139

hatch01 commented:
Just to be curious and learn more on nix, why don't you put a nix-shell of flake to run the update script.
Because it is a python script, it has dependencies at least on python3.

hatch01 commented:
I tested to update flutter to 3.16.5 using this script and got this error:
```shell
$ python3 update.py 
Flutter version(e.g. 3.13.8): 3.16.5
Engine commit(e.g. 767d8c75e898091b925519803830fc2721658d07): 3f3e560236539b7e2702f5ac790b2a4691b32d49
Dart version(e.g. 3.1.4): 3.2.3
Get artifect hashes? (y/n): y
error:
… <borked>

at «none»:0: (source not available)

… from call site

at «string»:1:6:

1| with import ./. {}; callPackage /tmp/tmp70wfpl69 {}
|      ^

error: function 'anonymous lambda' called without required argument 'callPackage'

at /home/eymeric/tmp/nixpkgs/pkgs/development/compilers/flutter/default.nix:1:1:

1| { callPackage, fetchzip, fetchFromGitHub, dart, lib, stdenv }:
| ^
2| let
{}
error:
… <borked>

at «none»:0: (source not available)

… from call site

at «string»:1:6:

1| with import ./. {}; callPackage /tmp/tmpk7m894mp {}
|      ^

error: function 'anonymous lambda' called without required argument 'callPackage'

at /home/eymeric/tmp/nixpkgs/pkgs/development/compilers/flutter/default.nix:1:1:

1| { callPackage, fetchzip, fetchFromGitHub, dart, lib, stdenv }:
| ^
2| let
Traceback (most recent call last):
  File "/home/eymeric/tmp/nixpkgs/pkgs/development/compilers/flutter/update.py", line 185, in <module>
    main()
  File "/home/eymeric/tmp/nixpkgs/pkgs/development/compilers/flutter/update.py", line 167, in main
    dart_hash = get_dart_hashes(dart_version).strip().replace('\n', '\n  ')
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/eymeric/tmp/nixpkgs/pkgs/development/compilers/flutter/update.py", line 138, in get_dart_hashes
    result_dict[platform] = pattern.findall(stderr)[0]
                            ~~~~~~~~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range
```

FlafyDev commented:
> I tested to update flutter to 3.16.5 using this script and got this error:

The script is designed to run from the root of the nixpkgs repo.  Maybe I should change it..

> Just to be curious and learn more on nix, why don't you put a nix-shell of flake to run the update script.
Because it is a python script, it has dependencies at least on python3.

I think the convention in nixpkgs is to just make a script and reference it in passthru.updateScript
I don't really see why use a flake. No need for any inputs or a special shell. And not everyone uses flakes.

FlafyDev commented:
The update script is now supposed to run from `pkgs/development/compilers/flutter`.
Also, it receives `--version`, `--dart-version`, `--engine-hash`, and `--artifect-hashes` instead of asking the user directly(with `input()`).

hatch01 commented:
Would it be possible to also get the hashes for the wrapper.nix ?

adamcstephens commented:
An updateScript should determine what the latest version is, not expect the user to provide that information. As written there is no point in adding to passthru.updateScript since it will error without those arguments. 

Scripts are generally written to be run from the root of nixpkgs, though ideally they should support being run from anywhere.

Have you read the section in the nixpkgs manual and reviewed other scripts?

FlafyDev committed:
flutter: Move to a structure more fitting for multiple versions

FlafyDev commented:
> Have you read the section in the nixpkgs manual and reviewed other scripts?

I couldn't find much info about update scripts and I missed the entry in the manual. I've read it now 

> An updateScript should determine what the latest version is, not expect the user to provide that information. As written there is no point in adding to passthru.updateScript since it will error without those arguments.

I made the script get the latest stable flutter version so there's now no need to provide any arguments for it to work.

> Scripts are generally written to be run from the root of nixpkgs, though ideally they should support being run from anywhere.

I also made the script automatically get the root of Nixpkgs, so it can now run anywhere as long as it's inside Nixpkgs.


Also, currently the script outputs the updated Nix code to the terminal so whoever updates Flutter can copy and paste it to Nixpkgs code. So since the script doesn't automatically update Nixpkgs code for now, should I keep the line `updateScript = ./update.py;`, anyway? Or should I remove it until it can?

mkg20001 commented:
> Also, currently the script outputs the updated Nix code to the terminal so whoever updates Flutter can copy and paste it to Nixpkgs code. So since the script doesn't automatically update Nixpkgs code for now, should I keep the line updateScript = ./update.py;, anyway? Or should I remove it until it can?

Update scripts are expected to modify nixpkgs code. For example r-ryantm will run it and then expect to be able to create a PR based on the changes. So better leave it out.

You can look at other scripts, there are helper command lines available for modifying hashes. Otherwise you can just replace the entire block of code that you generate. Or you can put the nix code into a single file which is imported and overwritten/managed entierly by the script (this would be the easiest and that's what some packages do).

FlafyDev commented:
> Update scripts are expected to modify nixpkgs code. For example r-ryantm will run it and then expect to be able to create a PR based on the changes. So better leave it out.
> 
> You can look at other scripts, there are helper command lines available for modifying hashes. Otherwise you can just replace the entire block of code that you generate. Or you can put the nix code into a single file which is imported and overwritten/managed entierly by the script (this would be the easiest and that's what some packages do).


Okay. So in order to make the script update Nixpkgs code I'd need to also implement the script to get "pubspecLockFile", "vendorHash", and "depsListFile". Which aren't needed in #263345, so it will just be extra work once it's merged.. 
And it also requires changing the structure of `pkgs/development/compilers/flutter/*` a bit to make it easier to replace/add Nix code from the script.

So because of that I think I'll wait on this PR until #263345 is merged. Then I'll continue working on this..

mkg20001 commented:
> So because of that I think I'll wait on this PR until https://github.com/NixOS/nixpkgs/pull/263345 is merged. Then I'll continue working on this..

OK.

Btw, since I'm not sure if/when hacker1024 will have time would you want to rebase the PR and apply your suggestions yourself for https://github.com/NixOS/nixpkgs/pull/263345 in a branch? I will then force-push that into the PR and merge that.

FlafyDev commented:
> > So because of that I think I'll wait on this PR until #263345 is merged. Then I'll continue working on this..
> 
> OK.
> 
> Btw, since I'm not sure if/when hacker1024 will have time would you want to rebase the PR and apply your suggestions yourself for #263345 in a branch? I will then force-push that into the PR and merge that.
> 
> (Edit: I'm not well versed in the new flutter code and honestly would prefer somebody else take care of the rebase)

Yeah, I will work on that. thanks

Henry-Hiles commented:
Fixes #272977

mkg20001 commented:
@FlafyDev The other PR is now merged, please rebase.

FlafyDev committed:
flutter: Add update script

FlafyDev committed:
fluffychat: Lock to flutter 3.13

FlafyDev committed:
localsend: Lock to flutter 3.13

hacker1024 commented:
I just tried to do the 3.16 upgrade myself, but the compiler seems to be crashing in macOS. Is that happening here as well?

FlafyDev commented:
> I just tried to do the 3.16 upgrade myself, but the compiler seems to be crashing on macOS when building flutter_tools. Is that happening here as well?

I don't have a darwin device to test this on.. It works fine for me on x86_64. 
Didn't test arm either

FlafyDev commented:
As per ofborg it seems flutter builds fine on x86_64 and aarch linux, but not x86_64 darwin.

https://logs.ofborg.org/?attempt_id=7e311264-7b67-4dbd-94df-aab01533f941&key=nixos%2Fnixpkgs.276050

FlafyDev commented:
I just tried building `flutter` from `nixpkgs-unstable` on my x86_64-darwin VM(without the binary cache, `--option substitute false`). And I got a compile error for flutter tools version 3.13.8.

The binary cache somehow has a successful build of flutter_tools on unstable.
But when compiling it locally it doesn't work.. Or at least that's from my experience on a VM.

I guess that means this PR wasn't the one that introduced this problem.. Maybe an issue in the sandboxing for darwin? Or am I missing something?

hacker1024 commented:
The issue does seem to be due to some other change in Nixpkgs then. Who can we ask about this? The longer we wait, the trickier it will be to find what broke this.

It's probably worth making an issue on the Dart SDK tracker as well, as the compiler shouldn't really be crashing like that anyway...

FlafyDev commented:
Result of `nixpkgs-review pr 276050` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>expidus.file-manager</li>
    <li>expidus.file-manager.pubcache</li>
    <li>firmware-updater</li>
    <li>firmware-updater.pubcache</li>
    <li>fluffychat</li>
    <li>fluffychat.pubcache</li>
    <li>flutter</li>
    <li>flutter313</li>
    <li>hover</li>
    <li>intiface-central</li>
    <li>intiface-central.pubcache</li>
    <li>localsend</li>
    <li>localsend.pubcache</li>
    <li>yubioath-flutter</li>
    <li>yubioath-flutter.pubcache</li>
  </ul>
</details>

FlafyDev commented:
> The issue does seem to be due to some other change then. Who can we ask about this? The longer we wait, the trickier it will be to find what broke this.
> 
> I don't think it's related to the Nix version, as I've barely touched my macOS installation since I last used it to get hashes.
> 
> It's probably worth making an issue on the Dart SDK tracker as well, as the compiler shouldn't really be crashing like that anyway...

I'm not sure what to do about it.. I am not very good with macos.
I tried building flutter tools locally on macos(not through nix) and I encountered weird errors such as 
```
Exception: Could not find directory at /Users/flafy/flutter/bin/cache/dart-sdk/bin/resources/devtools
```
when trying to compile with `flutter.dart`.

Though not including macos not working(which I'm not sure is related to the PR), I think this PR is ready for review.

FlafyDev commented:
Result of `nixpkgs-review pr 276050` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>expidus.file-manager</li>
    <li>expidus.file-manager.pubcache</li>
    <li>firmware-updater</li>
    <li>firmware-updater.pubcache</li>
    <li>fluffychat</li>
    <li>fluffychat.pubcache</li>
    <li>flutter</li>
    <li>flutter313</li>
    <li>hover</li>
    <li>intiface-central</li>
    <li>intiface-central.pubcache</li>
    <li>localsend</li>
    <li>localsend.pubcache</li>
    <li>yubioath-flutter</li>
    <li>yubioath-flutter.pubcache</li>
  </ul>
</details>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3256

lucasew commented on pkgs/development/compilers/flutter/update.py:
```diff
@@ -0,0 +1,359 @@
+    while True:
+        line = process.stderr.readline()
+        if not line:
+            break
```
`line = line.decode('utf-8')` so you decode all once

FlafyDev replied:
True
Though I ended up just using `text=True`

lucasew commented on pkgs/development/compilers/flutter/update.py:
```diff
@@ -0,0 +1,359 @@
+    for (platform, code) in code.items():
+        stderr = nix_build_to_fail(code % (dartVersion))
+
+        pattern = re.compile(r"got:\s+(.+?)\n")
```
You can replicate Nix hashes using the Python stdlib

Example: https://github.com/lucasew/playground/blob/master/piton/fetch-sri-sha256.py

Note that it only works with single files.

FlafyDev replied:
that doesn't work for me with fetchzip. gives a different hash.

How necessary is that?

mkg20001 replied:
is fetchzip being run with unpack flag? then it won't replicate them 1:1

lucasew commented:
Result of `nixpkgs-review pr 276050` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>expidus.file-manager</li>
    <li>expidus.file-manager.pubcache</li>
    <li>firmware-updater</li>
    <li>firmware-updater.pubcache</li>
    <li>fluffychat</li>
    <li>fluffychat.pubcache</li>
    <li>flutter</li>
    <li>flutter313</li>
    <li>hover</li>
    <li>intiface-central</li>
    <li>intiface-central.pubcache</li>
    <li>localsend</li>
    <li>localsend.pubcache</li>
    <li>yubioath-flutter</li>
    <li>yubioath-flutter.pubcache</li>
  </ul>
</details>

lucasew commented on pkgs/development/compilers/flutter/update.py:
```diff
@@ -0,0 +1,359 @@
+    regex = r"FINE: Contents:\n.*Generated by pub(.|\n)*?ERR :"
+    matches = re.search(regex, stderr)
+    pubspec_lock_yaml = '\n'.join(
+        [line[6:] for line in matches.group().strip().splitlines()[1:-1]])
```
Can we actually count on this line slice behaviour?

Another approach may be do this inside a fixed output derivation with empty hash on purpose then print a marker in the derivation before and after cat the lockfile and then extract the lockfile from nix logs. The derivation will obvously fail but will output the lockfile in the logs, then just find the markers in the logs and extract what is in between the markers.

lucasew replied:
Advantages of this approach
- Can run on remote builders.
- Harder to misparse, with a marker with exact size of dashes (`echo '----------'` before and after cat pubspec.lock ) it's very unlikely to get it wrong.
- Every user using that will have Nix anyway.
- No stuff will be written in their home folder (`~/.pub`), not even the `result` symlink in this case because the derivation will fail.

FlafyDev replied:
good idea! done

lucasew reviewed (commented):
BTW, about the approach, why not generate one single JSON file for each flutter version with the lockfile, SDK hashes and maybe the GitHub code instead of a few Nix files? buildFlutterPackage already requires one to use `lib.importJSON` to bring the lockfile to build stuff.

Henry-Hiles reviewed (approved):
Lgtm other than the mac issue aand what lucasaw said.

lucasew commented:
> I just tried building `flutter` from `nixpkgs-unstable` on my x86_64-darwin VM(without the binary cache, `--option substitute false`). And I got a compile error for flutter tools version 3.13.8.
> 
> The binary cache somehow has a successful build of flutter_tools on unstable. But when compiling it locally it doesn't work.. Or at least that's from my experience on a VM.
> 
> I guess that means this PR wasn't the one that introduced this problem.. Maybe an issue in the sandboxing for darwin? Or am I missing something?

I think someone forgot to update the hash of the darwin tools. It's not hard to miss this as a nix build of flutter on linux will not evaluate the source derivations to darwin. Ideally flutter could passthru a derivation to join it all just to make sure it touches all the source derivations so the hashes are updated accordingly. Can be something like `passthru.tests.srcs` maybe.

FlafyDev commented:
> BTW, about the approach, why not generate one single JSON file for each flutter version with the lockfile, SDK hashes and maybe the GitHub code instead of a few Nix files? buildFlutterPackage already requires one to use lib.importJSON to bring the lockfile to build stuff.

oh yeah. I think it would be a good idea to merge `default.nix` and `pubspec.lock.json` into single json files(maybe named `data.json` or `hashes.json`? Not sure). This will save the update script from trying to convert Python objects to Nix attrsets.

We'd still need to keep the `version/vX_XX/<files>` hierarchy since each version might have a `patches` directory for version specific patches.

> I think someone forgot to update the hash of the darwin tools. It's not hard to miss this as a nix build of flutter on linux will not evaluate the source derivations to darwin. Ideally flutter could passthru a derivation to join it all just to make sure it touches all the source derivations so the hashes are updated accordingly. Can be something like passthru.tests.srcs maybe.

What do you mean by the hash of darwin tools? do you mean like the artifact hashes for darwin? because these are only used after building flutter_tools.

And in general it seems that every dart package is failing on darwin, not related to flutter.

lucasew commented:
> What do you mean by the hash of darwin tools? do you mean like the artifact hashes for darwin? because these are only used after building flutter_tools.

Yes

> And in general it seems that every dart package is failing on darwin, not related to flutter.

Then focus in make sure that the artifacts and other related hashes are correct but focus on fixing build on Linux. Fix Darwin if you have the resources (Mac). If someone on Mac really cares then they can submit a follow up PR later.

FlafyDev commented:
Result of `nixpkgs-review pr 276050` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>yubioath-flutter</li>
    <li>yubioath-flutter.pubcache</li>
  </ul>
</details>
<details>
  <summary>13 packages built:</summary>
  <ul>
    <li>expidus.file-manager</li>
    <li>expidus.file-manager.pubcache</li>
    <li>firmware-updater</li>
    <li>firmware-updater.pubcache</li>
    <li>fluffychat</li>
    <li>fluffychat.pubcache</li>
    <li>flutter</li>
    <li>flutter313</li>
    <li>hover</li>
    <li>intiface-central</li>
    <li>intiface-central.pubcache</li>
    <li>localsend</li>
    <li>localsend.pubcache</li>
  </ul>
</details>

FlafyDev commented:
`yubioath-flutter` fails due to `pngquant` failing on upstream on aarch64-linux (tested with `nix build github:nixos/nixpkgs#pngquant --system aarch64-linux`)

FlafyDev commented:
Result of `nixpkgs-review pr 276050` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>expidus.file-manager</li>
    <li>expidus.file-manager.pubcache</li>
    <li>firmware-updater</li>
    <li>firmware-updater.pubcache</li>
    <li>fluffychat</li>
    <li>fluffychat.pubcache</li>
    <li>flutter</li>
    <li>flutter313</li>
    <li>hover</li>
    <li>intiface-central</li>
    <li>intiface-central.pubcache</li>
    <li>localsend</li>
    <li>localsend.pubcache</li>
    <li>yubioath-flutter</li>
    <li>yubioath-flutter.pubcache</li>
  </ul>
</details>

lucasew commented:
Result of `nixpkgs-review pr 276050` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>expidus.file-manager</li>
    <li>expidus.file-manager.pubcache</li>
    <li>firmware-updater</li>
    <li>firmware-updater.pubcache</li>
    <li>fluffychat</li>
    <li>fluffychat.pubcache</li>
    <li>flutter</li>
    <li>flutter313</li>
    <li>hover</li>
    <li>intiface-central</li>
    <li>intiface-central.pubcache</li>
    <li>localsend</li>
    <li>localsend.pubcache</li>
    <li>yubioath-flutter</li>
    <li>yubioath-flutter.pubcache</li>
  </ul>
</details>

FlafyDev commented:
Can this be merged? Is there anything left to do?

mkg20001 commented on pkgs/development/compilers/flutter/update.py:
```diff
@@ -0,0 +1,386 @@
+    os.fsync(temp.fileno())
+
+    process = subprocess.Popen(
+        ["nix", "build", "--impure", "--print-out-paths", "--no-link", "--expr",
```
this depends on experimental features. please add the necesarry flags for enabling the features in this command.

mkg20001 commented on pkgs/development/compilers/flutter/update.py:
```diff
@@ -0,0 +1,386 @@
+
+def get_artifact_hashes(flutter_compact_version):
+    code = """
+        {
```
if possible put the code in a file, to keep this script short

FlafyDev replied:
I don't want to clutter the flutter directory with update script files. Should I make a `../compilers/flutter/update` directory and put the update script there along with the nix files?

mkg20001 replied:
Yes, sounds reasonable

FlafyDev replied:
updated. what do you think?

FlafyDev commented:
Result of `nixpkgs-review pr 276050` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>expidus.file-manager</li>
    <li>expidus.file-manager.pubcache</li>
    <li>firmware-updater</li>
    <li>firmware-updater.pubcache</li>
    <li>fluffychat</li>
    <li>fluffychat.pubcache</li>
    <li>flutter</li>
    <li>flutter313</li>
    <li>hover</li>
    <li>intiface-central</li>
    <li>intiface-central.pubcache</li>
    <li>localsend</li>
    <li>localsend.pubcache</li>
    <li>yubioath-flutter</li>
    <li>yubioath-flutter.pubcache</li>
  </ul>
</details>

hacker1024 commented on pkgs/development/compilers/flutter/patches/disable-auto-update.patch:
```diff
@@ -1,16 +1,3 @@
-     exit 1
-   fi
- 
--  upgrade_flutter 7< "$PROG_NAME"
```
Ah, `internal.sh` is no longer used. I wonder if it's worth deleting the file entirely so that if anything _does_ try to use it it will fail in a cleaner way?

FlafyDev replied:
Hmm I'm not sure. But I think deleting it is a good move. 
Though, I don't think this is really related to this PR I think. I just had to edit the patch so I could update to 3.16.7 and I noticed we are never using `internal.sh` anyway.

hacker1024 replied:
In that case, could we update the patch instead for now rather than removing it? I doubt it matters too much but after #280875 everything scares me...

```
--- a/bin/internal/shared.sh
+++ b/bin/internal/shared.sh
@@ -245,7 +245,7 @@ function shared::execute() {
   # and will corrupt each others' downloads.
   #
   # SHARED_NAME itself is prepared by the caller script.
-  upgrade_flutter 7< "$SHARED_NAME"
+  # upgrade_flutter 7< "$SHARED_NAME"
 
   BIN_NAME="$(basename "$PROG_NAME")"
   case "$BIN_NAME" in
```

FlafyDev replied:
Done. Each flutter version has it's own patch for that file now.

FlafyDev commented on pkgs/development/compilers/flutter/update/update.py:
```diff
@@ -0,0 +1,340 @@
+
+    stderr = nix_build_to_fail(code)
+
+    pattern = re.compile(r"For full logs, run '+(.+?)'\.\n")
```
You need to call `nix log <derivation's .drv file>. Which I don't know of another way to get aside from getting it from the user messages.

Does this comment for a stable api also apply to the `got: <hash>` messages this script is parsing? Since I don't know of another way of getting it and it's what other update script/tools also do.

I've ended up using a derivation over calling pub locally because:
- Nothing will be written to the user's home folder in `~/.pub`
- I didn't want to have to copy and paste the entire flutter repository to somewhere local from nix store and deal with permissions

Not sure if these are good reasons, but if we can get `pubspec.lock` from the tarball I think that is the preferable way.. What do you think?

hacker1024 replied:
> You need to call `nix log <derivation's .drv file>. Which I don't know of another way to get aside from getting it from the user messages.

`nix-instantiate` should do it - instead of `nix-build`, use `nix-instantiate`, grab the `.drv` path, build it with `nix-store --realise`, and check the logs with `nix log`.

> Does this comment for a stable api also apply to the got: <hash> messages this script is parsing? Since I don't know of another way of getting it and it's what other update script/tools also do.

I suppose it does, but I don't know any other way either.

> Not sure if these are good reasons, but if we can get pubspec.lock from the tarball I think that is the preferable way.. What do you think?

The tarball is more of a last resort as it has the potential to download quite a lot.

I'm thinking, when you use a blank hash and it spits out the real one, it still keeps that output in the store, right? Because when you rebuild with the right hash it's instant. So I wonder if we can make the FOD output the `pubspec.lock`. and use the hash from the error message to obtain the true output path. Then we can just copy the `pubspec.lock` from there.

lucasew replied:
Doing in a derivation avoids the mess of filling `~/.pub` for example and forces it to download the dependencies.

BTW another, simpler, approach is the following:
- Get the derivation path using `nix eval .#package.drvPath`
- Build the derivation with `nix-store -r /nix/store/thatderivation.drv`
- Get the logs of the build using `nix log /nix/store/thatderivation.drv` then the rest is parse work basically.

FlafyDev replied:
Ahh I'm not sure which way is better.
For now I'll change to the @hacker1024 's method since we already do it this way to get Flutter's source.

hacker1024 commented:
I have fixed the macOS compilation issues in #280875, but some artifact hashes appear to be wrong. I am not entirely sure that the use of the version patch is enough to correctly set the artifact platform; it was never meant for that to begin with.

```
┏━ 2 Errors: 
┃ error: hash mismatch in fixed-output derivation '/nix/store/nlmys1hksch6y3mhcbd9wlmpigh216w0-flutter-artifacts-android-x86_64-darwin.drv':
┃          specified: sha256-j8jstEE1RsTVHJbq6f6We0An+CyJz9JH/YClyNA4mwg=
┃             got:    sha256-0FBI0CGMcxyttkzrdyjJlkGAjFd/yMuAQS3pDrNXZZw=
┃ error: hash mismatch in fixed-output derivation '/nix/store/9zmspxfx323k13y127wmj9rszsd9rmpn-flutter-artifacts-universal-x86_64-darwin.drv':
┃          specified: sha256-quSFKx7TZRJpK+4YDt5f9jwr7rZsSsaXMxhJ8vIcczQ=
┃             got:    sha256-Qwf12gMqrW5nDC9Is08oxWTbKMptRQRAIb58JETq3xA=
┣━ Dependency Graph:
┃    ┌─ ✔ flutter-artifacts-macos-x86_64-darwin ⏱ 38s
┃    ├─ ✔ flutter-artifacts-ios-x86_64-darwin ⏱ 56s
┃    │  ┌─ ⚠ flutter-artifacts-android-x86_64-darwin failed with hash mismatch after ⏱ 1m35s
┃    ├─ ⏸ flutter-artifacts-android-x86_64-darwin
┃    │  ┌─ ⚠ flutter-artifacts-universal-x86_64-darwin failed with hash mismatch after ⏱ 58s
┃    ├─ ⏸ flutter-artifacts-universal-x86_64-darwin
┃    │  ┌─ ✔ flutter-artifacts-web-x86_64-darwin ⏱ 33s
┃    ├─ ✔ flutter-artifacts-web-x86_64-darwin ⏱ 27s
┃ ┌─ ⏸ flutter-cache-dir
┃ ⏸ flutter-wrapped-3.16.7-sdk-links
┣━━━ Builds         
┗━ ∑ ⏵ 0 │ ✔ 4 │ ⏸ 4 │ ⚠ Exited after 2 build failures at 22:41:35 after 1m49s
```

FlafyDev commented:
> I have fixed the macOS compilation issues in #280875, but some artifact hashes appear to be wrong. I am not entirely sure that the use of the version patch is enough to correctly set the artifact platform; it was never meant for that to begin with.
> 
> ```
> ┏━ 2 Errors: 
> ┃ error: hash mismatch in fixed-output derivation '/nix/store/nlmys1hksch6y3mhcbd9wlmpigh216w0-flutter-artifacts-android-x86_64-darwin.drv':
> ┃          specified: sha256-j8jstEE1RsTVHJbq6f6We0An+CyJz9JH/YClyNA4mwg=
> ┃             got:    sha256-0FBI0CGMcxyttkzrdyjJlkGAjFd/yMuAQS3pDrNXZZw=
> ┃ error: hash mismatch in fixed-output derivation '/nix/store/9zmspxfx323k13y127wmj9rszsd9rmpn-flutter-artifacts-universal-x86_64-darwin.drv':
> ┃          specified: sha256-quSFKx7TZRJpK+4YDt5f9jwr7rZsSsaXMxhJ8vIcczQ=
> ┃             got:    sha256-Qwf12gMqrW5nDC9Is08oxWTbKMptRQRAIb58JETq3xA=
> ┣━ Dependency Graph:
> ┃    ┌─ ✔ flutter-artifacts-macos-x86_64-darwin ⏱ 38s
> ┃    ├─ ✔ flutter-artifacts-ios-x86_64-darwin ⏱ 56s
> ┃    │  ┌─ ⚠ flutter-artifacts-android-x86_64-darwin failed with hash mismatch after ⏱ 1m35s
> ┃    ├─ ⏸ flutter-artifacts-android-x86_64-darwin
> ┃    │  ┌─ ⚠ flutter-artifacts-universal-x86_64-darwin failed with hash mismatch after ⏱ 58s
> ┃    ├─ ⏸ flutter-artifacts-universal-x86_64-darwin
> ┃    │  ┌─ ✔ flutter-artifacts-web-x86_64-darwin ⏱ 33s
> ┃    ├─ ✔ flutter-artifacts-web-x86_64-darwin ⏱ 27s
> ┃ ┌─ ⏸ flutter-cache-dir
> ┃ ⏸ flutter-wrapped-3.16.7-sdk-links
> ┣━━━ Builds         
> ┗━ ∑ ⏵ 0 │ ✔ 4 │ ⏸ 4 │ ⚠ Exited after 2 build failures at 22:41:35 after 1m49s
> ```
> 
> ```
> --- /nix/store/ik96y9b15zcaj13acq1xgszrxx7jwp1j-flutter-artifacts-universal-x86_64-darwin
> +++ /nix/store/i6zj3ll0jd5m2yzvqpyz0ihc3ddalsay-flutter-artifacts-universal-x86_64-darwin
> │   --- /nix/store/ik96y9b15zcaj13acq1xgszrxx7jwp1j-flutter-artifacts-universal-x86_64-darwin/artifacts
> ├── +++ /nix/store/i6zj3ll0jd5m2yzvqpyz0ihc3ddalsay-flutter-artifacts-universal-x86_64-darwin/artifacts
> │ │   --- /nix/store/ik96y9b15zcaj13acq1xgszrxx7jwp1j-flutter-artifacts-universal-x86_64-darwin/artifacts/engine
> │ ├── +++ /nix/store/i6zj3ll0jd5m2yzvqpyz0ihc3ddalsay-flutter-artifacts-universal-x86_64-darwin/artifacts/engine
> │ │ ├── file list
> │ │ │ @@ -1,2 +1,2 @@
> │ │ │  common
> │ │ │ -linux-x64
> │ │ │ +darwin-x64
> ```

managed to solve it. the issue is that operatingSystem in flutter_tools needs to be spoofed as well..
By adding this patch

```
diff --git a/packages/flutter_tools/lib/src/base/platform.dart b/packages/flutter_tools/lib/src/base/platform.dart
index 45da89ad4c..2d79dbaece 100644
--- a/packages/flutter_tools/lib/src/base/platform.dart
+++ b/packages/flutter_tools/lib/src/base/platform.dart
@@ -132,7 +132,7 @@ class LocalPlatform extends Platform {
   String get pathSeparator => io.Platform.pathSeparator;

   @override
-  String get operatingSystem => io.Platform.operatingSystem;
+  String get operatingSystem => io.Platform.environment['NIX_FLUTTER_OPERATING_SYSTEM'] ?? io.Platform.operatingSystem;

   @override
   String get operatingSystemVersion => io.Platform.operatingSystemVersion;
```

And adding  
```
  export NIX_FLUTTER_OPERATING_SYSTEM=${if systemPlatform == ''x86_64-darwin'' then ''macos'' else ''linux''}
```
to `fetch-artifacts.nix`, you get all the correct hashes.
I confirmed the hashes on flutter 3.13.8.

What do you think of this method? It may be a bit hacky and we need to add another patch.. But I don't know of any other way.

hacker1024 commented:
> > I have fixed the macOS compilation issues in #280875, but some artifact hashes appear to be wrong. I am not entirely sure that the use of the version patch is enough to correctly set the artifact platform; it was never meant for that to begin with.
> > ```
> > ┏━ 2 Errors: 
> > ┃ error: hash mismatch in fixed-output derivation '/nix/store/nlmys1hksch6y3mhcbd9wlmpigh216w0-flutter-artifacts-android-x86_64-darwin.drv':
> > ┃          specified: sha256-j8jstEE1RsTVHJbq6f6We0An+CyJz9JH/YClyNA4mwg=
> > ┃             got:    sha256-0FBI0CGMcxyttkzrdyjJlkGAjFd/yMuAQS3pDrNXZZw=
> > ┃ error: hash mismatch in fixed-output derivation '/nix/store/9zmspxfx323k13y127wmj9rszsd9rmpn-flutter-artifacts-universal-x86_64-darwin.drv':
> > ┃          specified: sha256-quSFKx7TZRJpK+4YDt5f9jwr7rZsSsaXMxhJ8vIcczQ=
> > ┃             got:    sha256-Qwf12gMqrW5nDC9Is08oxWTbKMptRQRAIb58JETq3xA=
> > ┣━ Dependency Graph:
> > ┃    ┌─ ✔ flutter-artifacts-macos-x86_64-darwin ⏱ 38s
> > ┃    ├─ ✔ flutter-artifacts-ios-x86_64-darwin ⏱ 56s
> > ┃    │  ┌─ ⚠ flutter-artifacts-android-x86_64-darwin failed with hash mismatch after ⏱ 1m35s
> > ┃    ├─ ⏸ flutter-artifacts-android-x86_64-darwin
> > ┃    │  ┌─ ⚠ flutter-artifacts-universal-x86_64-darwin failed with hash mismatch after ⏱ 58s
> > ┃    ├─ ⏸ flutter-artifacts-universal-x86_64-darwin
> > ┃    │  ┌─ ✔ flutter-artifacts-web-x86_64-darwin ⏱ 33s
> > ┃    ├─ ✔ flutter-artifacts-web-x86_64-darwin ⏱ 27s
> > ┃ ┌─ ⏸ flutter-cache-dir
> > ┃ ⏸ flutter-wrapped-3.16.7-sdk-links
> > ┣━━━ Builds         
> > ┗━ ∑ ⏵ 0 │ ✔ 4 │ ⏸ 4 │ ⚠ Exited after 2 build failures at 22:41:35 after 1m49s
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > --- /nix/store/ik96y9b15zcaj13acq1xgszrxx7jwp1j-flutter-artifacts-universal-x86_64-darwin
> > +++ /nix/store/i6zj3ll0jd5m2yzvqpyz0ihc3ddalsay-flutter-artifacts-universal-x86_64-darwin
> > │   --- /nix/store/ik96y9b15zcaj13acq1xgszrxx7jwp1j-flutter-artifacts-universal-x86_64-darwin/artifacts
> > ├── +++ /nix/store/i6zj3ll0jd5m2yzvqpyz0ihc3ddalsay-flutter-artifacts-universal-x86_64-darwin/artifacts
> > │ │   --- /nix/store/ik96y9b15zcaj13acq1xgszrxx7jwp1j-flutter-artifacts-universal-x86_64-darwin/artifacts/engine
> > │ ├── +++ /nix/store/i6zj3ll0jd5m2yzvqpyz0ihc3ddalsay-flutter-artifacts-universal-x86_64-darwin/artifacts/engine
> > │ │ ├── file list
> > │ │ │ @@ -1,2 +1,2 @@
> > │ │ │  common
> > │ │ │ -linux-x64
> > │ │ │ +darwin-x64
> > ```
> 
> managed to solve it. the issue is that operatingSystem in flutter_tools needs to be spoofed as well.. By adding this patch
> 
> ```
> diff --git a/packages/flutter_tools/lib/src/base/platform.dart b/packages/flutter_tools/lib/src/base/platform.dart
> index 45da89ad4c..2d79dbaece 100644
> --- a/packages/flutter_tools/lib/src/base/platform.dart
> +++ b/packages/flutter_tools/lib/src/base/platform.dart
> @@ -132,7 +132,7 @@ class LocalPlatform extends Platform {
>    String get pathSeparator => io.Platform.pathSeparator;
> 
>    @override
> -  String get operatingSystem => io.Platform.operatingSystem;
> +  String get operatingSystem => io.Platform.environment['NIX_FLUTTER_OPERATING_SYSTEM'] ?? io.Platform.operatingSystem;
> 
>    @override
>    String get operatingSystemVersion => io.Platform.operatingSystemVersion;
> ```
> 
> And adding
> 
> ```
>   export NIX_FLUTTER_OPERATING_SYSTEM=${if systemPlatform == ''x86_64-darwin'' then ''macos'' else ''linux''}
> ```
> 
> to `fetch-artifacts.nix`, you get all the correct hashes. I confirmed the hashes on flutter 3.13.8.
> 
> What do you think of this method? It may be a bit hacky and we need to add another patch.. But I don't know of any other way.

Sounds good to me - this would let us maintain Apple Silicon support as well.

That environment variable name is a little vague though - maybe something like `NIX_FLUTTER_ARTIFACT_BUILD_PLATFORM`? Or maybe something less lengthy, names are hard 😶

FlafyDev committed:
flutter: Make it possible to override `operatingSystem` internally

Makes it possible to get artifact hashes for other system platforms.

Useful for an update script.

FlafyDev committed:
flutter: 3.13.8 -> 3.16.7

FlafyDev commented on pkgs/development/compilers/flutter/update/update.py:
```diff
@@ -0,0 +1,340 @@
+            "--impure",
+            "--print-out-paths",
+            "--no-link",
+            "--experimental-features", "nix-command flakes",
```
True!
Change it to nix-build in both functions

FlafyDev commented:
> 
> Sounds good to me - this would let us maintain Apple Silicon support as well.
> 
> That environment variable name is a little vague though - maybe something like `NIX_FLUTTER_ARTIFACT_BUILD_PLATFORM`? Or maybe something less lengthy, names are hard 😶

I kept `NIX_FLUTTER_HOST_PLATFORM` and `NIX_FLUTTER_OPERATING_SYSTEM` as they are because I want them to tell exactly what they do in flutter_tools:

 `NIX_FLUTTER_HOST_PLATFORM` = changes host platform
 `NIX_FLUTTER_OPERATING_SYSTEM` = changes operating system

and we have 2 variables that control the artifact platform, can't name them both `NIX_FLUTTER_ARTIFACT_BUILD_PLATFORM`

lucasew commented:
Result of `nixpkgs-review pr 276050` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>expidus.file-manager</li>
    <li>expidus.file-manager.pubcache</li>
    <li>firmware-updater</li>
    <li>firmware-updater.pubcache</li>
    <li>fluffychat</li>
    <li>fluffychat.pubcache</li>
    <li>flutter</li>
    <li>flutter313</li>
    <li>hover</li>
    <li>intiface-central</li>
    <li>intiface-central.pubcache</li>
    <li>localsend</li>
    <li>localsend.pubcache</li>
    <li>yubioath-flutter</li>
    <li>yubioath-flutter.pubcache</li>
  </ul>
</details>

hacker1024 reviewed (changes requested):
This is looking pretty good now! Just a minor nitpick about the ordering in `all-packages.nix`.

Tested building on x86_64-linux for Linux and Web, Android is still broken (#260278), but that can be fixed in a later PR.

hacker1024 commented:
Result of `nixpkgs-review pr 276050` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>expidus.file-manager</li>
    <li>expidus.file-manager.pubcache</li>
    <li>firmware-updater</li>
    <li>firmware-updater.pubcache</li>
    <li>fluffychat</li>
    <li>fluffychat.pubcache</li>
    <li>flutter (flutter316 ,flutterPackages.stable ,flutterPackages.v3_16)</li>
    <li>flutter313 (flutterPackages.v3_13)</li>
    <li>intiface-central</li>
    <li>intiface-central.pubcache</li>
    <li>localsend</li>
    <li>localsend.pubcache</li>
    <li>yubioath-flutter</li>
    <li>yubioath-flutter.pubcache</li>
  </ul>
</details>

FlafyDev commented on pkgs/top-level/all-packages.nix:
```diff
@@ -15805,10 +15805,10 @@ with pkgs;
-  flutter = flutterPackages.wrapFlutter flutter-unwrapped;
+  flutterPackages = recurseIntoAttrs (callPackage ../development/compilers/flutter { });
+  flutter = flutterPackages.stable;
+  flutter316 = flutterPackages.v3_16;
+  flutter313 = flutterPackages.v3_13;
```
I think there is inconsistency in the sorting method I used, which is why the order is different for us.

I think I fixed it now.

The change
```patch
-    versions = sorted(versions, key=lambda x: int(
-        x.split('_')[0]), reverse=True)
+    versions = sorted(versions, key=lambda x: (
+        int(x.split('_')[0]), int(x.split('_')[1])), reverse=True)
```

Can you test again?

FlafyDev commented:
Result of `nixpkgs-review pr 276050` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>expidus.file-manager</li>
    <li>expidus.file-manager.pubcache</li>
    <li>firmware-updater</li>
    <li>firmware-updater.pubcache</li>
    <li>fluffychat</li>
    <li>fluffychat.pubcache</li>
    <li>flutter</li>
    <li>flutter313</li>
    <li>hover</li>
    <li>intiface-central</li>
    <li>intiface-central.pubcache</li>
    <li>localsend</li>
    <li>localsend.pubcache</li>
    <li>yubioath-flutter</li>
    <li>yubioath-flutter.pubcache</li>
  </ul>
</details>

FlafyDev commented:
Result of `nixpkgs-review pr 276050` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>expidus.file-manager</li>
    <li>expidus.file-manager.pubcache</li>
    <li>firmware-updater</li>
    <li>firmware-updater.pubcache</li>
    <li>fluffychat</li>
    <li>fluffychat.pubcache</li>
    <li>flutter</li>
    <li>flutter313</li>
    <li>hover</li>
    <li>intiface-central</li>
    <li>intiface-central.pubcache</li>
    <li>localsend</li>
    <li>localsend.pubcache</li>
    <li>yubioath-flutter</li>
    <li>yubioath-flutter.pubcache</li>
  </ul>
</details>
